util/kernel_names.h

```diff
@@ -0,0 +1,139 @@
+ *           there are more of them.
+ *           This function uses discarded statements to map input type into the
+ *           return type we may use for the kernel name template parameters.
+ *           Note: it does not provide different kernel names for the type
+ *           aliases, so we may have the same kernel name
+ *             - for `int` and `int32_t`, or
+ *             - for `size_t` and `uint64_t`, or
+ *             - for any other possible type alias
+ *           on any specific SYCL implementation. We may say it provides adapter
+ *           aliases for the type aliases.
+ */
+template <typename T>
+constexpr auto adapter_helper() {
+  // Fixed-width types are optional and implementation-defined
+#ifdef INT8_MAX
+  if constexpr (std::is_same_v<T, std::int8_t>) {
+    return adapter_int8_t{};
+  } else
+#endif
+#ifdef INT16_MAX
+      if constexpr (std::is_same_v<T, std::int16_t>) {
+    return adapter_int16_t{};
+  } else
+#endif
+#ifdef INT32_MAX
+      if constexpr (std::is_same_v<T, std::int32_t>) {
+    return adapter_int32_t{};
+  } else
+#endif
+#ifdef INT64_MAX
+      if constexpr (std::is_same_v<T, std::int64_t>) {
+    return adapter_int64_t{};
+  } else
+#endif
+#ifdef UINT8_MAX
+      if constexpr (std::is_same_v<T, std::uint8_t>) {
+    return adapter_uint8_t{};
+  } else
+#endif
+#ifdef UINT16_MAX
+      if constexpr (std::is_same_v<T, std::uint16_t>) {
+    return adapter_uint16_t{};
+  } else
+#endif
+#ifdef UINT32_MAX
+      if constexpr (std::is_same_v<T, std::uint32_t>) {
+    return adapter_uint32_t{};
+  } else
+#endif
+#ifdef UINT64_MAX
+      if constexpr (std::is_same_v<T, std::uint64_t>) {
+    return adapter_uint64_t{};
+  } else
+#endif
+      if constexpr (std::is_same_v<T, std::byte>) {
+    // std::byte is a scoped enum according to the C++17
+    return adapter_byte{};
+  } else if constexpr (std::is_same_v<T, std::size_t>) {
+    return adapter_size_t{};
+  } else {
+    return adapter_core<T>{};
+  }
+}
+
+};  // namespace details
+
+/** @brief Safe wrapper for types to use within kernel names
+ *  @details According to C++17 [dcl.typedef] type alias does not introduce a
+ *           new type. So we are free to provide adapter for types we cannot use
+ *           for kernel names directly. For example, for any type we cannot
+ *           forward-declare:
+ *               struct outer {
+ *                 struct inner {};
+ *               };
+ *           we are unable to have
+ *               using T = outer::inner;
+ *               single_task<kernel_name<T>>,
+ *           but we are able to have
+ *               using T = typename kernel_name::adapter<outer::inner>::type
+ *               single_task<kernel_name<T>>
+ *           in case we provide a proper template specialization for adapter.
+ *
+ *           Test developer is free to provide template specialization for any
+ *           test-specific type.
```

Comment:
This is an interesting design document that deserves to appear close to the top of this file.

Reply:
I was thinking about it in terms of doxygen documentation section for `struct adapter`. So every time we modify it or every time we end up near the `adapter_t` declaration using IDE or full-text search we would see the appropriate documentation nearby.
AFAIK I cannot move a `struct adapter` up because of C++ restrictions. @keryell Should I duplicate design notice in the header comment or move it entirely?

Reply:
Done by 28e77dc09abc162c641cf412f084239d32410c0f